haystack_experimental/components/tools/openapi/openapi_tool.py

```diff
@@ -0,0 +1,162 @@
+
+logger = logging.getLogger(__name__)
+
+
+@component
+class OpenAPITool:
+    """
+    The OpenAPITool calls a RESTful endpoint of an OpenAPI service using payloads generated from human instructions.
+
+    Here is an example of how to use the OpenAPITool component to scrape a URL using the FireCrawl API:
+
+    ```python
+    from haystack.dataclasses import ChatMessage
+    from haystack_experimental.components.tools.openapi import OpenAPITool, LLMProvider
+
+    tool = OpenAPITool(generator_api=LLMProvider.OPENAI,
+                       generator_api_params={"model":"gpt-3.5-turbo"},
+                       tool_spec="https://raw.githubusercontent.com/mendableai/firecrawl/main/apps/api/openapi.json",
+                       tool_credentials="<your-tool-token>")
+
+    results = tool.run(messages=[ChatMessage.from_user("Scrape URL: https://news.ycombinator.com/")])
+    print(results)
+    ```
+
+    Similarly, you can use the OpenAPITool component to invoke **any** OpenAPI service/tool by providing the OpenAPI
+    specification and credentials.
+    """
+
+    def __init__(
+        self,
+        generator_api: LLMProvider,
+        generator_api_params: Dict[str, Any],
+        tool_spec: Optional[Union[str, Path]] = None,
+        tool_credentials: Optional[str] = None,
+    ):
+        """
+        Initialize the OpenAPITool component.
+
+        :param generator_api: The API provider for the chat generator.
+        :param generator_api_params: Parameters to pass for the chat generator creation.
+        :param tool_spec: OpenAPI specification for the tool/service. This can be a URL, a local file path, or
+        an OpenAPI service specification provided as a string.
+        :param tool_credentials: Credentials for the tool/service.
+        """
+        self.generator_api = generator_api
+        self.chat_generator = self._init_generator(generator_api, generator_api_params)
+        self.config_openapi: Optional[ClientConfiguration] = None
+        self.open_api_service: Optional[OpenAPIServiceClient] = None
+        if tool_spec:
+            self.config_openapi = ClientConfiguration(
+                openapi_spec=tool_spec,
+                credentials=tool_credentials,
+                llm_provider=generator_api
+            )
+            self.open_api_service = OpenAPIServiceClient(self.config_openapi)
+
+    @component.output_types(service_response=List[ChatMessage])
+    def run(
+        self,
+        messages: List[ChatMessage],
+        fc_generator_kwargs: Optional[Dict[str, Any]] = None,
+        tool_spec: Optional[Union[str, Path]] = None,
+        tool_credentials: Optional[str] = None,
+    ) -> Dict[str, List[ChatMessage]]:
+        """
+        Invokes the underlying OpenAPI service/tool with the function calling payload generated by the chat generator.
+
+        :param messages: List of ChatMessages to generate function calling payload (e.g. human instructions).
```

Comment:
We should mention what the expected format is seeing how we make assumptions about the list and its contents.